flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/abilities/source/SourceAbilitySpec.java

```diff
@@ -56,4 +57,12 @@
      */
     @JsonIgnore
     Optional<RowType> getProducedType();
+
+    /**
+     * When Filter,Limit,Partition,etc. are pushed into {@link DynamicTableSource}, the source will
+     * add a digest to record this behavior.
+     *
+     * @param exprStrs The expressions need to be added into the extra digest.
+     */
+    String getExtraDigests(Optional<List<String>> exprStrs);
```

Comment:
I think this should return an array or a list. As it stands now, we're pulling the formatting concern into this method that was previously separated.